inst/templates/ad_adlb.R

```diff
@@ -0,0 +1,276 @@
+)
+
+
+# ---- Derivations ----
+
+# Get list of ADSL vars required for derivations
+adsl_vars <- vars(TRTSDT, TRTEDT, TRT01A, TRT01P)
+
+adlb <- lb %>%
+
+  # Join ADSL with LB (need TRTSDT for ADY derivation)
+  derive_vars_merged(
+    dataset_add = adsl,
+    new_vars = adsl_vars,
+    by_vars = vars(STUDYID, USUBJID)
+  ) %>%
+
+  # Calculate ADT, ADY
+  derive_vars_dt(
+    new_vars_prefix = "A",
+    dtc = LBDTC,
+    flag_imputation = FALSE
+  ) %>%
+
+  derive_var_ady(reference_date = TRTSDT, date = ADT)
+
+adlb <- adlb %>%
+
+  # Add PARAMCD PARAM and PARAMN - from LOOK-UP table
+  # Replace with PARAMCD lookup function
+  derive_vars_merged(
+    dataset_add = param_lookup,
+    new_vars = vars(PARAMCD, PARAM, PARAMN),
+    by_vars = vars(LBTESTCD)
+  ) %>%
+
+  # Calculate PARCAT1 AVAL AVALC ANRLO ANRHI
+  mutate(
+    PARCAT1 = LBCAT,
+    AVAL  = LBSTRESN,
+    AVALC = LBSTRESC,
+    ANRLO = LBSTNRLO,
+    ANRHI = LBSTNRHI
+  )
+
+# get visit info
+adlb <- adlb %>%
+
+  # Derive Timing
+  mutate(
+    AVISIT = case_when(
+      str_detect(VISIT, "SCREEN") ~ "Baseline",
+      !is.na(VISIT) ~ str_to_title(VISIT),
+      TRUE ~ NA_character_
+    ),
+    AVISITN = case_when(
+      AVISIT == "Baseline" ~ 0,
+      !is.na(VISITNUM) ~ VISITNUM
+    )
+  )
+
+adlb <- adlb %>%
+
+  # Calculate ONTRTFL
+  derive_var_ontrtfl(
+    start_date = ADT,
+    ref_start_date = TRTSDT,
+    ref_end_date = TRTEDT,
+    filter_pre_timepoint = AVISIT == "Baseline"
+  )
+
+# Calculate ANRIND : requires the reference ranges ANRLO, ANRHI
+adlb <- adlb %>%
+  derive_var_anrind()
+
+# Derive baseline flags
+adlb <- adlb %>%
+  # Calculate BASETYPE
+  mutate(
+    BASETYPE = "LAST"
+  ) %>%
+
+  # Calculate ABLFL
+  derive_var_extreme_flag(
+    by_vars = vars(STUDYID, USUBJID, BASETYPE, PARAMCD),
+    order = vars(ADT, VISITNUM, LBSEQ),
+    new_var = ABLFL,
+    mode = "last",
+    filter = (!is.na(AVAL) & ADT <= TRTSDT & !is.na(BASETYPE))
+  )
+
+# Derive baseline information
+adlb <- adlb %>%
+
+  # Calculate BASE
+  derive_var_base(
+    by_vars = vars(STUDYID, USUBJID, PARAMCD, BASETYPE),
+    source_var = AVAL,
+    new_var = BASE
+  ) %>%
+
+  # Calculate BASEC
+  derive_var_base(
+    by_vars = vars(STUDYID, USUBJID, PARAMCD, BASETYPE),
+    source_var = AVALC,
+    new_var = BASEC
+  ) %>%
+
+  # Calculate BNRIND
+  derive_var_base(
+    by_vars = vars(STUDYID, USUBJID, PARAMCD, BASETYPE),
+    source_var = ANRIND,
+    new_var = BNRIND
+  ) %>%
+
+  # Calculate CHG
+  derive_var_chg() %>%
+
+  # Calculate PCHG
+  derive_var_pchg()
+
+
+# Add RATIO derivations with new function??
+# R2BASE, R2ANRLO R2ANRHI
+adlb <- adlb %>%
+  mutate(
+    R2BASE = if_else(
+      !is.na(AVAL) & !is.na(BASE) & BASE != 0, AVAL / BASE, NA_real_
+    ),
+    R2ANRLO = if_else(
+      !is.na(AVAL) & !is.na(ANRLO) & ANRLO != 0, AVAL / ANRLO, NA_real_
+    ),
+    R2ANRHI = if_else(
+      !is.na(AVAL) & !is.na(ANRHI) & ANRHI != 0, AVAL / ANRLO, NA_real_
```

Comment:
Should R2ANRHI have AVAL / ANRHI in the denominator?

Reply:
Cobbling some pseduo-code together for ration function.  Do you think we need any additional warnings messages/error messages around what is missing as we require 4 variables to be present in the dataset.  I guess assert_data_frame will give us that info  

```
derive_vars_analysis_ratio <- function(dataset) {
  AVAL <- assert_symbol(enquo(AVAL))
  BASE <- assert_symbol(enquo(BASE))
  ANRLO <- assert_symbol(enquo(ANRLO))
  ANRHI <- assert_symbol(enquo(ANRHI))
  assert_data_frame(dataset, required_vars = quo_c(AVAL, BASE, ANRLO, ANRHI))

  dataset %>% mutate(
    R2BASE = if_else(
      !is.na(AVAL) & !is.na(BASE) & BASE != 0, AVAL / BASE, NA_real_
    ),
    R2ANRLO = if_else(
      !is.na(AVAL) & !is.na(ANRLO) & ANRLO != 0, AVAL / ANRLO, NA_real_
    ),
    R2ANRHI = if_else(
      !is.na(AVAL) & !is.na(ANRHI) & ANRHI != 0, AVAL / ANRHI, NA_real_
    )
  )
}
```

Reply:
I think that check is fine, I guess they should all be numeric? Not sure if we check for that in other functions. I would assume user can pass in what derivations they want so if they only wanted R2BASE, then no need to check for ANRLO ANRHI. So check would be dependent on which derivation thery requested.